### ISSUE TITLE
Cannot read properties of undefined (reading 'dropDisabled' / 'dragDisabled')

### ISSUE THREAD

**Issue by DManstrator** [vaadin/flow-components#2627]
### Description

I get the errors above after updating from VAADIN 21.0.4 to VAADIN 21.0.5. I only noticed it now after updating to VAADIN 22 which added console errors to the VAADIN / Spring debug tools. With changing versions I could determine VAADIN 21.0.5 as the update since when those errors start to appear.

There errors appear already when loading the site, not just when trying to drag and drop elements.

Full errors:
> There seems to be an error in Vaadin Grid: Cannot read properties of undefined (reading 'dropDisabled') Please submit an issue to https://github.com/vaadin/flow-components/issues/new/choose

> There seems to be an error in Vaadin Grid: Cannot read properties of undefined (reading 'dragDisabled') Please submit an issue to https://github.com/vaadin/flow-components/issues/new/choose

Something interesting I noticed: I always got `items-1` error logs. If the grid had 23 elements I got 22 errors in the console.

### Expected outcome

I don't expect those errors to occur.

### Actual outcome

Those errors occur (as stupid as it sounds).

### Live Demo (optional)

Only errors are logged, I don't think screenshots / a video is needed.

### Minimal reproducible example

```java
Grid<CustomObject> grid = new Grid<>();

// Configure dragging
grid.setRowsDraggable(true);
grid.setDragFilter(node -> true);
grid.addDragStartListener(event -> {});
grid.addDragEndListener(event -> {});

// Configure dropping
grid.setDropMode(GridDropMode.ON_TOP);
grid.addDropListener(event -> {});

// This happens later, perhaps this is the reason for the error.
grid.setItems(list);
```

### Steps to reproduce

Nothing special treatement is required, simply open the site which runs that code.

### Environment

- VAADIN 21.0.5 +
- Windows 10 Version 21H1 (x64)

### Browsers Affected

Removed untested browsers to avoid confusion.

- [x] Chrome
- [x] Edge

**Comment by tomivirkki** [comment]
Hi @DManstrator 

I tried your snippet with the latest Vaadin 22 version (22.0.5) and didn't get any errors. Here's the full example I used:

```java
package com.example.application.views.helloworld;

import java.util.ArrayList;
import java.util.List;
import com.vaadin.flow.component.grid.Grid;
import com.vaadin.flow.component.grid.dnd.GridDropMode;
import com.vaadin.flow.component.html.Div;

import com.vaadin.flow.router.Route;

@Route(value = "hello-world")
public class HelloWorldView extends Div  {


    public HelloWorldView() {
        List<CustomObject> list = new ArrayList<>();
        list.add(new CustomObject("John"));
        list.add(new CustomObject("Jane"));


        // Snippet from the issue description
        Grid<CustomObject> grid = new Grid<>();

        // Configure dragging
        grid.setRowsDraggable(true);
        grid.setDragFilter(node -> true);
        grid.addDragStartListener(event -> {});
        grid.addDragEndListener(event -> {});

        // Configure dropping
        grid.setDropMode(GridDropMode.ON_TOP);
        grid.addDropListener(event -> {});

        // This happens later, perhaps this is the reason for the error.
        grid.setItems(list);

        add(grid);
    }

    public static class CustomObject {
        private String name;

        public CustomObject(String name) {
            this.name = name;
        }

        public String getName() {
            return name;
        }

        public void setName(String name) {
            this.name = name;
        }
    }
}

```

**Comment by DManstrator** [comment]
Sorry for the late reply, I sadly had no time to further investigate my issue so far. Our codebase regarding this is (for some reason) large which is why it takes long to properly investigate the issue I face. More attributes are set on the Grid between setting the drop and drag listeners and setting the items, so perhaps this is the reason your example works. The items are also set in a different method which is called automatically from VAADIN from what I can tell so far. I just saw that not setting the listeners before values being present got rid of the errors but I have to dig into our code to be able to give proper information.

TL; DR: This is not resolved yet and I will investigate this further when I find the time for it, please don't close the issue due to inactivity. Thanks.

**Comment by DManstrator** [comment]
Hello Tomi,

I finally found the time to further investigate my issue. Despite the codebase is really big and complex, I found out that the bug itself is luckly simple to reproduce.

The items can be set much later (not even within the same request) in a `reload` method which is just called when the component is getting loaded. The drag / drop is configured in the constructor but the constructor is called as soon as the main component gets loaded.

I don't know how easy that is to understand, so I will explain the architecture behind a bit.

We have a main component with a header. In the header, there are sub components linked as Tabs. The tabs are defined like this:

```java
// custom interface we created
protected TabComponent[] createTabComponents() {
    return new TabComponent[] {
            // ...
            new HelloWorldView(),
           // ...
    };
}

// Reduced code, should only display the important part.
public interface TabComponent extends Serializable {

    void reload();

}
```

I adjusted your `HelloWorldView` to match our architecture:

```java
public class HelloWorldView extends Div implements TabComponent {

    private final Grid<CustomObject> grid = new Grid<>();

    public HelloWorldView() {
        // [...]

        grid.addColumn(CustomObject::getName).setHeader("Name");

        // Configure dragging
        grid.setRowsDraggable(true);
        grid.setDragFilter(node -> true);
        grid.addDragStartListener(event -> {});
        grid.addDragEndListener(event -> {});

        // Configure dropping
        grid.setDropMode(GridDropMode.ON_TOP);
        grid.addDropListener(event -> {});

        add(grid);
        setSizeFull();

        // Note that the items are not set here!
    }

    @Override
    public void reload() {
        final List<CustomObject> customObjects = Arrays.asList(new CustomObject("John"), new CustomObject("Jane"));
        grid.setItems(customObjects);
    }

}
```

The `createTabComponents` method is invoked when the main component is loaded (to properly display the tabs) so each constructor (including the one of the `HelloWorldView`) is getting invoked. There, the drop and drag is configured.

But only when the `HelloWorldView` is getting loaded, its `reload` method is invoked which finally sets the items.

And if this doesn't happen in the same cycle, the errors mentioned above are shown in the console. At least this is what I saw. When I started the application while targeting the `HelloWorldView`, I got no errors. If another tab was targeted when starting the application and I opened the `HelloWorldView` later, the errors have appeared in the console.

I hope what I wrote is understandable and that I didn't cut off too many things from the code. As said, it is quite big and making a minimal reproducible example is pretty hard.

Please let me know if you need more context to this and I will do my best to provide you that!

Kind regards,

Daniel

**Comment by sissbruecker** [comment]
@DManstrator From your example it's not clear when `reload` is called. I tried this again with the following setup, using a button to add the grid and set the data in a second roundtrip, it's still not reproducable.
```java
public class HelloWorldView extends Div {
    private final TabContent tabContent;

    public HelloWorldView() {
        // Initialize tab content
        tabContent = new TabContent();

        // Button for adding tab content, and loading
        Button addTab = new Button("Add tab", e -> {
            add(tabContent);
            tabContent.reload();
        });
        add(addTab);
    }

    private static class TabContent extends Div {
        private final Grid<CustomObject> grid = new Grid<>();

        public TabContent() {
            grid.addColumn(CustomObject::getName).setHeader("Name");

            // Configure dragging
            grid.setRowsDraggable(true);
            grid.setDragFilter(node -> true);
            grid.addDragStartListener(event -> {});
            grid.addDragEndListener(event -> {});

            // Configure dropping
            grid.setDropMode(GridDropMode.ON_TOP);
            grid.addDropListener(event -> {});

            add(grid);
            setSizeFull();
        }

        public void reload() {
            final List<CustomObject> customObjects = Arrays.asList(new CustomObject("John"), new CustomObject("Jane"));
            grid.setItems(customObjects);
        }
    }

    private static class CustomObject {
        private final String name;

        public CustomObject(String name) {
            this.name = name;
        }

        public String getName() {
            return name;
        }
    }
}
```

Please also note that the Vaadin version you mentioned in the OP (v21) is not supported anymore, so this needs to be reproducable for either v22 or v23. I attached my demo project for reproduction, you can check if you can manage to reproduce the error there: [grid-drag-undefined.zip](https://github.com/vaadin/flow-components/files/8287911/grid-drag-undefined.zip)

**Comment by DManstrator** [comment]
Hello Sasha,

I'm sorry that I was so unspecific regarding when the reload method is getting called.

Also, thank you very much for that your example project.

Regarding the VAADIN version: It first appeared in 21.0.5 but it is still present in latest VAADIN 22 version. I also managed to reproduce the error with the VAADIN version of the example project you gave me (23.0.1).

### What I changed:
- I added our `TabComponent` interface to the project and added more required methods I left out when commenting last time.
- In addition, I successfully abstracted our architecture as good as possible and added it to the example project.

For this, I only needed to adjust the contents of the `views` folder (`src\main\java\com\example\application\views`).

I hope this helps you. Thank you again for everything you did so far.

The adjusted example project: [grid-drag-undefined_bug-reproducible.zip](https://github.com/vaadin/flow-components/files/8323999/grid-drag-undefined_bug-reproducible.zip)

**Comment by sissbruecker** [comment]
Thanks, I can reproduce it now. The clue seems to be toggling the visibility of the grid and setting items in the same roundtrip, while having drag and drop configured. A simple reproduction is:
```java
Grid<String> grid = new Grid<>();
grid.addColumn(ValueProvider.identity()).setHeader("Item");
grid.setRowsDraggable(true);
grid.setVisible(false);

Button show = new Button("Show", e -> {
    grid.setItems("A", "B", "C");
    grid.setVisible(true);
});

add(grid, show);
```

Interestingly I'm getting a somewhat different error message (with the simple and with the full reproduction), but still get (n - 1) error messages:
```
There seems to be an error in Vaadin Grid:
rowData.item is undefined
Please submit an issue to https://github.com/vaadin/flow-components/issues/new/choose 4 simple:6:224
    error simple:6
    tryCatchWrapper FlowBootstrap.js:67
    _filterDragAndDrop vaadin-grid-drag-and-drop-mixin.js:378
    filterDragAndDrop vaadin-grid-drag-and-drop-mixin.js:367
    filterDragAndDrop vaadin-grid-drag-and-drop-mixin.js:366
    _dragDropAccessChanged vaadin-grid-drag-and-drop-mixin.js:354
```

**Comment by DManstrator** [comment]
Interesting.. I double checked it just now, thinking that VAADIN 23 could have made a change to it. But for me, the full reproduction (v23.0.1) still gives me the same error message as reported in this issue.

![image](https://user-images.githubusercontent.com/7572614/159654169-3dc92a1d-03da-4952-9924-ed9929d443d5.png)

I didn't adjust any dependency from the example project you gave me so I cannot really explain it. The only thing I noticed is that I get much less details regarding the error than you got.

I think it is very unlikely but does perhaps the node version or the host OS make a difference here?

When starting the program, I noticed that VAADIN uses a custom node version brought by itself.
> 2022-03-23 09:10:47.736  INFO 17132 --- [  restartedMain] c.v.flow.server.frontend.FrontendTools   : Project node version 16.13.1 is older than 16.14.0. Using node from C:/Users/userName/.vaadin.

Also I'm running the latest Windows 10.

**Comment by sissbruecker** [comment]
Setting severity to minor, as this seems to be more of an annoyance and does not prevent using the grid afterwards. If this actually breaks the grid in other ways, please comment, and we can reconsider the severity.

**Comment by stefanuebe** [comment]
A potential workaround to prevent those message is to delegate the setItems / data provider refresh to a second roundtrip. Might have some impact on the loading time, but prevents the error messages from popping up

```
grid.getElement().executeJs("return").then(v -> grid.getDataProvider().refreshAll());
```

Sample code to test: 

```
grid.setVisible(false);

Button show = new Button("Show/Refresh", e -> {
    if (!grid.isVisible()) {
        grid.setVisible(true);
        grid.getElement().executeJs("return").then(v -> grid.getDataProvider().refreshAll());
    } else {
        grid.getDataProvider().refreshAll();
    }
});
add(grid, show);
```
Without the `...executeJs("return").then(...)` the message appears as usual. When using the return expression, it will not.

**Comment by circlesmiler** [comment]
I have the same problem.

**Comment by apklein** [comment]
I started receiving this error when I upgraded from 24.0.6 to 24.1.0. My code is large and I haven't clue already about what's happening.

**Comment by knoobie** [comment]
I have the same problem with a Drag-n-Drop enabled Grid within a Dialog which is really annoying. Every single time I open the Dialog.

**Comment by knoobie** [comment]
I can confirm it's fixed with 24.1.6+ (I was still on an earlier version back then).  Took me some time to find the correct commits for 24.1.6, #5250 sadly has no link to #5254 with the information when it was fixed
